### PR TITLE
Restore `DocBuilder::new()` to avoid breaking API change

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

- Closes https://github.com/apache/datafusion/issues/13764

## Rationale for this change

I would like to help people upgrade DataFusion versions with as much help as possible. This means following the [deprecation guidelines](https://github.com/apache/datafusion/blob/main/README.md) 

Changes to `DocBuilder` introduced a breaking API change that could be confusing to some during upgrade
- https://github.com/apache/datafusion/pull/13575

## What changes are included in this PR?
1. Restore `DocumentationBuilder::new`, and mark it  `#[deprecated]``, following the [deprecation guidelines](https://github.com/apache/datafusion/blob/main/README.md) 
2. Make `DocumentationBuilder::new_with_defaults`  new to help users find the new API
3. Add `Default` for `DocSection` to make it easier for users to bash out their own docs
4. Improve some other docs

### CODE REVIEW

**Comment by alamb** [comment]
@Omega359 or @comphead  I wonder if you have time to review this change prior to the 44 release?

File: datafusion/sql/src/statement.rs [hunk 1]
```diff
@@ -62,8 +62,8 @@ use sqlparser::ast::{
     Assignment, AssignmentTarget, ColumnDef, CreateIndex, CreateTable,
     CreateTableOptions, Delete, DescribeAlias, Expr as SQLExpr, FromTable, Ident, Insert,
     ObjectName, ObjectType, OneOrManyWithParens, Query, SchemaName, SetExpr,
-    ShowCreateObject, Statement, TableConstraint, TableFactor, TableWithJoins,
-    TransactionMode, UnaryOperator, Value,
+    ShowCreateObject, ShowStatementFilter, Statement, TableConstraint, TableFactor,
```
Comment (Omega359) [inline datafusion/sql/src/statement.rs:65]: cargo fmt drive by change?

**Review by Omega359 — APPROVED** [review]
LGTM

File: datafusion/sql/src/statement.rs (hunk 1 above)
Comment (alamb) [reply datafusion/sql/src/statement.rs:65]: I think it was something I needed to to do iix the build. I'll merge up and it should go away

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb

**Comment by alamb** [comment]
Thank you @comphead and @Omega359

**Pull request merged**